Introduce `clang-tidy` and optimize code

hebasto opened the pull request:
This PR:
- introduces the `-DLibmultiprocess_ENABLE_CLANG_TIDY` configuration option, which allows to "run clang-tidy with the compiler"
- fixes some `clang-tidy`'s warnings, which are easy-to-review and useful on their own

Sjors commented:
I got one warning during `make`, same as on master, building on Ubuntu 22.10 with clang 15.0.6:

```
[ 75%] Built target multiprocess
[ 87%] Building CXX object CMakeFiles/mpgen.dir/src/mp/gen.cpp.o
/home/sjors/dev/libmultiprocess/src/mp/gen.cpp: In function ‘void Generate(kj::StringPtr, kj::StringPtr, kj::StringPtr, kj::ArrayPtr<const kj::StringPtr>)’:
/home/sjors/dev/libmultiprocess/src/mp/gen.cpp:179:44: warning: ‘capnp::ParsedSchema capnp::SchemaParser::parseDiskFile(kj::StringPtr, kj::StringPtr, kj::ArrayPtr<const kj::StringPtr>) const’ is deprecated [-Wdeprecated-declarations]
  179 |     auto file_schema = parser.parseDiskFile(src_file, src_file, import_paths);
      |                        ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/sjors/dev/libmultiprocess/src/mp/gen.cpp:9:
/usr/include/capnp/schema-parser.h:101:16: note: declared here
  101 |   ParsedSchema parseDiskFile(kj::StringPtr displayName, kj::StringPtr diskPath,
      |                ^~~~~~~~~~~~~
[100%] Linking CXX executable mpgen
[100%] Built target mpgen
```

`make check` fails entirely though, while on master it works fine.

```
[ 73%] Building CXX object test/CMakeFiles/mptest.dir/mp/test/foo.capnp.proxy-server.c++.o
In file included from /home/sjors/dev/libmultiprocess/include/mp/proxy.h:8,
                 from /home/sjors/dev/libmultiprocess/build/test/mp/test/foo.capnp.proxy.h:8,
                 from /home/sjors/dev/libmultiprocess/build/test/mp/test/foo.capnp.proxy-types.h:6,
                 from /home/sjors/dev/libmultiprocess/build/test/mp/test/foo.capnp.proxy-server.c++:3:
/home/sjors/dev/libmultiprocess/include/mp/util.h: In instantiation of ‘mp::AsyncCallable<typename std::remove_reference<_T
…
/home/sjors/dev/libmultiprocess/build/test/mp/test/foo.capnp.proxy-server.c++:23:24:   required from here
/home/sjors/dev/libmultiprocess/include/mp/util.h:349:24: error: no matching function for call to ‘forward(kj::CaptureByMove<kj::CaptureByMove<mp::PassField<Accessor<foo_fields::Context, 17>,
…
ooInterface::CallbackResults> > > >&)’
  349 |     return std::forward(callable);
      |            ~~~~~~~~~~~~^~~~~~~~~~
In file included from /usr/include/c++/12/bits/stl_pair.h:61,
                 from /usr/include/c++/12/bits/stl_algobase.h:64,
                 from /usr/include/c++/12/bits/stl_tree.h:63,
                 from /usr/include/c++/12/map:60,
                 from /home/sjors/dev/libmultiprocess/test/mp/test/foo.h:8,
                 from /home/sjors/dev/libmultiprocess/build/test/mp/test/foo.capnp.proxy.h:7:
/usr/include/c++/12/bits/move.h:77:5: note: candidate: ‘template<class _Tp> constexpr _Tp&& std::forward(typename remove_reference<_Tp>::type&)’
   77 |     forward(typename std::remove_reference<_Tp>::type& __t) noexcept
```

[full log](https://gist.github.com/Sjors/3acea3fe33871e33e9daf5d6eff4e630)

hebasto commented:
@Sjors 

Thank you for your comment.

> `make check` fails entirely though, while on master it works fine.

The PR has been reworked now.

Sjors commented:
`make check` works with ceb3a4b8a0d5d04d7a8573a3dca4f71b4a8b04da

ryanofsky commented:
re: https://github.com/chaincodelabs/libmultiprocess/pull/83#issuecomment-1426695225

Note: warning about deprecated `parseDiskFile` call is reported as a standalone issue in #39

ryanofsky commented on CMakeLists.txt:
```diff
@@ -8,6 +8,15 @@ project("Libmultiprocess" CXX)
 set(CMAKE_CXX_STANDARD 17)
 set(CMAKE_CXX_STANDARD_REQUIRED YES)
 
+option(MULTIPROCESS_RUN_CLANG_TIDY "Run clang-tidy with the compiler." OFF)
```
In commit "Add option to run `clang-tidy` with compiler" (f3b2bfcbe7ecca087d75e96df1496687f8ff4c0a)

Would it make sense to call the option `ENABLE_CLANG_TIDY` instead of `MULTIPROCESS_RUN_CLANG_TIDY`? I don't see a reason to prefix option names with the project name, and the other options don't do this

hebasto replied:
The goal of prefixing is to minimize risk if option name clash when this library is a subproject of another project. For example, https://github.com/bitcoin-core/secp256k1/pull/1113.

ryanofsky replied:
Ok, I didn't think about this being used as a subdirectory inside another project. Could we then prefix project options with the project name like `Libmultiprocess_ENABLE_CLANG_TIDY`. I do also prefer `ENABLE` to `RUN` for the cmake option name, because at the `CMakeLists.txt` level, this isn't adding more build steps, just setting another option.

hebasto replied:
Thanks! [Updated](https://github.com/chaincodelabs/libmultiprocess/pull/83#issuecomment-1430948349).

ryanofsky reviewed (commented):
Confirmed ceb3a4b8a0d5d04d7a8573a3dca4f71b4a8b04da is working and runs clang but I didn't review the individual commits yet.

I also noticed one new warning is shown now (using clang-tidy 14.0.6)

```c++
include/mp/proxy-types.h:162:85: warning: the parameter 'perhaps' is copied for each invocation but only used as a const reference; consider making it a const reference [performance-unnecessary-value-param]
                            .then([&server, invoke, req](kj::Maybe<Thread::Server&> perhaps) {
                                                                                    ^
                                                         const                     &
```

ryanofsky commented on test/mp/test/foo.h:
```diff
@@ -48,7 +48,7 @@ class FooImplementation
     int callback(FooCallback& callback, int arg) { return callback.call(arg); }
     int callbackUnique(std::unique_ptr<FooCallback> callback, int arg) { return callback->call(arg); }
-    int callbackShared(std::shared_ptr<FooCallback> callback, int arg) { return callback->call(arg); }
+    int callbackShared(std::shared_ptr<FooCallback> callback, int arg) { return callback->call(arg); } // NOLINT(performance-unnecessary-value-param)
```
In commit "clang-tidy: Fix `performance-unnecessary-value-param` check" (0498ed86bc6d1a6bdbabf98f2e39468dc5c90823)

Note: I think it does make sense to suppress this lint warning than try to avoid it. Could potentially avoid it by changing this argument to `const shared_ptr` but there's currently no IPC serialization code to handle `const shared_ptr` only `shared_ptr`, so it does't compile. Also I don't think it would make sense to add const shared_ptr serialization code because most cases where you would pass `const shared<ptr>` it probably makes more sense to pass a plain const reference.

ryanofsky commented:
re: https://github.com/chaincodelabs/libmultiprocess/pull/83#pullrequestreview-1298380154

> I also noticed one new warning is shown now (using clang-tidy 14.0.6)

Following change fixes the warning:

```diff
--- a/include/mp/proxy-types.h
+++ b/include/mp/proxy-types.h
@@ -159,7 +159,7 @@ auto PassField(Priority<1>, TypeList<>, ServerContext& server_context, const Fn&
 
     auto thread_client = context_arg.getThread();
     return JoinPromises(server.m_context.connection->m_threads.getLocalServer(thread_client)
-                            .then([&server, invoke, req](kj::Maybe<Thread::Server&> perhaps) {
+                            .then([&server, invoke, req](const kj::Maybe<Thread::Server&>& perhaps) {
                                 KJ_IF_MAYBE(thread_server, perhaps)
                                 {
                                     const auto& thread = static_cast<ProxyServer<Thread>&>(*thread_server);

```

ryanofsky commented:
Code review ACK ceb3a4b8a0d5d04d7a8573a3dca4f71b4a8b04da, but I think I'd like to rename the cmake option as described https://github.com/chaincodelabs/libmultiprocess/pull/83#discussion_r1106345466. I noticed some other clang-tidy errors still occurred for me with clang-tidy 14.0.6 after this PR, so I separately merged fixes for these in https://github.com/chaincodelabs/libmultiprocess/pull/84

hebasto commented:
Updated 78bfa2a2812fe56b172ca4f16e7c332d0e03f103 -> d795e969dd7c913c937a4f97d78115072feb6cb4 ([pr83.02](https://github.com/hebasto/libmultiprocess/commits/pr83.02) -> [pr83.03](https://github.com/hebasto/libmultiprocess/commits/pr83.03)):

- rebased on top of the #84
- addressed @ryanofsky's [comment](https://github.com/chaincodelabs/libmultiprocess/pull/83#discussion_r1106443878)
- fixed `HeaderFilterRegex` for the `example` directory

ryanofsky reviewed (approved):
Code review ACK 594466ae975a8370bfed8f6de05bbcf92289dd8f. Changes since last review were changing option name, and updating HeaderFilterRegex, and rebasing

ryanofsky commented:
Merged now. Thank you for these nice changes and features!
